elastic/beats
# Add `elasticsearch/enrich` metricset

Resolves #14221.

This PR adds a new `enrich` metricset to the `elasticsearch` Metricbeat module. This metricset is responsible for collecting stats from the [`GET _enrich/_stats` Elasticsearch API](https://www.elastic.co/guide/en/elasticsearch/reference/master/enrich-stats-api.html). It's primarily intended to be used for Stack Monitoring, i.e. with the `xpack.enabled` flag set to `true` in the `elasticsearch` module configuration.

### Testing this PR
1. Build Metricbeat with this PR.
   ```
   cd $GOPATH/src/github.com/elastic/beats/metricbeat
   mage build
   ```
2. Enable the `elasticsearch-xpack` module. This module configuration contains the `enrich` metricset.
   ```
   ./metricbeat modules enable elasticsearch-xpack
   ```
3. Run Metricbeat.
   ```
   ./metricbeat -e
   ```

4. Follow the 6 steps at https://www.elastic.co/guide/en/elasticsearch/reference/7.x/ingest-enriching-data.html to ingest a document and have Elasticsearch enrich it. This will generate some enrich stats.

5. Give the `elasticsearch/enrich` metricset up to 10 seconds to collect enrich stats and index them.

5. Verify that the `.monitoring-es-*` indices contains enrich stats documents.
   ```
   # This will most likely return no hits, unless Metricbeat performed a collection
   # right as the enrich policy was still being executed.
   GET .monitoring-es-*/_search?q=type:enrich_executing_policy_stats&sort=timestamp:desc
   ```

   ```
   GET .monitoring-es-*/_search?q=type:enrich_coordinator_stats&sort=timestamp:desc
   ```
